File: go/fp-patterns/partial-application/partial-application.go
package main

import (
	"fmt"
)

func add(args ...int) int {
	var result int
	for _, v := range args {
		result += v
	}
	return result
}

// Вспомогательная функция для частичного применения с использованием обобщенной
// типизации
// Принимает в качестве аргумента функцию, аргументы которой нужно зафиксировать
// а так же список аргументов, которые нужно зафиксировать
// и возвращает функцию, которая готова принять новый список аргументов и
// вернуть результат вызова оригинальной функции со всеми переданными аргументами
// (зафиксированными + вновь переданными)
func partialApply[T any](fn func(args ...T) T, appliedArgs ...T) func(...T) T {
	return func(args ...T) T {
		return fn(append(appliedArgs, args...)...)
	}
}

func main() {
	// Простое частичное применение, посредством возврата результата вызова
	// функции с фиксированным аргументом
	addTwo := func(b int) int {
		return add(2, b)
	}
	result := addTwo(3)
	fmt.Println(result) // 5

	// Создание функции с зафиксированными аргументами посредством вспомогательной
	// функции
	addTwoPartiallyApplied := partialApply(add, 2, 3)
	result = addTwoPartiallyApplied(4)
	fmt.Println(result) // 9
}

File: typescript/fp-patterns/partial-application/partial-application.ts
// Простое частичное применение при помощи bind
const add = (a: number, b: number): number => {
  return a + b;
};

const addFive = add.bind(null, 5); // привяжем часть аргументов

console.log(addFive(3)); // 8
console.log(addFive(7)); // 12

// Частичное применения при помощи вспомогательной функции и apply
function apply(fn: Function, args: any[]) {
  return fn.apply(null, args);
}

const addFiveApply = apply.bind(null, add, [5]); // partial application of add function with [5] argument

console.log(addFive(3)); // 8
console.log(addFive(7)); // 12

// Обобщение частичного применения средствами Typescript
type PartiallyApplyArgs<TArgs extends any[], TReturn> = (
  ...args: TArgs
) => TReturn;

type PartialApply<TArgs extends any[], TReturn> = (
  fn: PartiallyApplyArgs<TArgs, TReturn>,
  ...args: TArgs
) => PartiallyApplyArgs<TArgs, TReturn>;

const partialApply: PartialApply<number[], number> =
  (fn, ...appliedArgs) =>
  (...args) =>
    fn(...appliedArgs, ...args);

const addPartial = partialApply(add, 1);
const result = addPartial(2);
console.log(result); // 3
